download.py

```diff
@@ -17,55 +20,102 @@
 @dataclass
 class DownloadResult:
     file_path: str  # target_file_path,  # TODO harmonize with dane-download-worker
-    download_time: float = -1  # time (secs) taken to receive data after request
-    mime_type: str = "unknown"  # download_data.get("mime_type", "unknown"),
+    mime_type: str
+    provenance: dict
+    download_time: float = -1  # time (ms) taken to receive data after request
+    error: str = ""
     content_length: int = -1  # download_data.get("content_length", -1),
 
 
-def download_uri(uri: str) -> Optional[DownloadResult]:
+def download_uri(uri: str) -> DownloadResult:
     logger.info(f"Trying to download {uri}")
     if validate_s3_uri(uri):
         logger.info("URI seems to be an s3 uri")
         return s3_download(uri)
-    return http_download(uri)
+    if validate_http_uri(uri):
+        logger.info("URI seems to be an HTTP URI")
+        return http_download(uri)
+    return DownloadResult(
+        uri, "", dict(), -1, "Input failure: URI is neither S3, nor HTTP"
+    )
 
 
 def http_download(url: str) -> DownloadResult:
     logger.info(f"Checking if {url} was already downloaded")
+    start_time = time.time()
+
+    provenance = {
+        "activity_name": "Input download",
+        "activity_description": "Downloads the input file from INPUT_URI",
+        "processing_time_ms": -1,
+        "start_time_unix": start_time,
+        "parameters": [],
+        "software_version": "",
+        "input_data": url,
+        "output_data": "",
+        "steps": [],
+    }
+
     fn = os.path.basename(urlparse(url).path)
-    input_file = os.path.join(input_file_dir, fn)
-    # TODO rethink how to return mime_type
-    _, extension = get_asset_info(input_file)
+    asset_id, extension = get_asset_info(fn)
```

Comment:
I disagree because the `task ID` should be reserved only for the API part. I want some sort of separation between the API part and the worker's actual processing part if that makes sense. Also, it would add more input for the worker part to process, whereas the `asset_id` is something that can simply be obtained from the input file's name.

I could indeed not use the asset_id and just use generic names like `input.mp4`, but if processing failed and the files were not deleted somehow, they would persist there and create issues with processing the next tasks, leading to a total worker failure. This can happen, for example, when trying to download a file but not having full access to the URL domain (like was the case when testing with input files from `modelhosting.beng.nl`).

Reply:
Ahh yes, very fair point about the task ID being just for the API part. I think what you are doing in practice is using the basename of the file, which makes sense. I just wouldn't call it `asset_id`, because a filename does not necesarily correspond to an id.